### ISSUE TITLE
Cold Staking contract

### ISSUE THREAD

**Issue by Dexaran** [EthereumCommonwealth/Roadmap#51]
@yograterol @RideSolo @yuriy77k 

I'd like to discuss the details of further Cold Staking contract implementation here.

At the moment we have 3 proposed variants:

1. [Dexaran's Cold Staking](https://github.com/EthereumCommonwealth/Cold-staking/tree/moonlight_3.1(testnet3)) ([description](https://docs.google.com/document/d/1LpRHzW7gGV1JfiXUQOmW-UEVF9TCNNWrYiItDuLgiGA/edit))

2. [Yuriy's Cold Staking](https://github.com/yuriy77k/Cold-staking/blob/master/ColdStaking.sol) ([description](https://docs.google.com/document/d/1uckDVbgAj6N6Nx6Sm8-CnL_kRJTT8MZRQfVEkT01ltM/edit))

3. [RideSolo's Cold Staking](https://github.com/RideSolo/Cold-staking/blob/master/ColdStaking.sol) ([description](https://drive.google.com/file/d/1fFtuQctoLtyEP-Lwlnt2x5pOuVWzZjn4/view?usp=sharing))

# Overview

### 1. Dexaran's Cold Staking

The first implementation of the Cold Staking protocol. Every action (new staker, reward claim) has a direct impact on the state of the contract which makes the final reward highly unpredictable.

Reward is computed at the moment of the claiming. Contract does not preserve the history of "staking weights". Reward depend on actual contract state and contract balance, thus it is independent of monetary policy. However, reward depends on claiming order (those, who claim their rewards first, earn more).

**Advantages: Has a working reference implementation.**

**Disadvantages: Each staker can affect others. Too unpredictable rewards.**

### 2. Yuriy's Cold Staking

This implementation relies on [TotalStakingWeight](https://github.com/yuriy77k/Cold-staking/blob/master/ColdStaking.sol#L127) variable which preserves the info about how much balance should be allocated for stakers who did not claim their rewards yet.

Stakers can not directly affect each other with their deposits/withdrawals/claims. Reward calculation is more predictable, reward only depend on the amount of staking CLO. The reward is independent of monetary policy.

**Advantages: Better reward calculation**

**Disadvantages: see `example 1`**

*Requires more testing.*

### 3. RideSolo's Cold Staking

Advanced reward calculation formula.
 However, this version is more complex compared to the previous one and it relies on [hardcoded block reward](https://github.com/RideSolo/Cold-staking/blob/master/ColdStaking.sol#L50) variable. Therefore it depends on monetary policy and it will be necessary to update the contract with a hardfork at each monetary policy change. Also, it is not compatible with The First Stake proposal.

**Advantages: Better reward calculation**

**Disadvantages: higher complexity, depend on monetary policy, incompatible with The First Stake**

# Implementation at Callisto Network

I'm in favor of implementing the (2) Yuriy's Cold Staking version. It can solve all pressing problems and the implementation is not overcomplicated.

I believe that it is necessary to take the following steps:

1. Deploy the contract at testnet, test it and make a final decision if it is ready or not.

2. Re-deploy the testnet and launch a testnet cold staking contract that will behave exactly like the final version (automatically receive block rewards, distribute staking rewards).

3. Perform a security audit.

4. Launch an official bug bounty.

5. (optional) Make final changes if necessary. If everything is fine then the contract is ready to be used. Update the whitepaper.

**Comment by RideSolo** [comment]
Hello @Dexaran, 
First of all, the proposal was to show an idea about how to handle the cold staking with a different angle so everyone can see that we have other possibility. 

I will try to comment and ask some questions about the disadvantages: 

- Higher complexity: mathematical proof can be more complex, but the implementation is simple.
- Incompatible with the first stake proposal : the proposed code was just to show the formula and how fair it can be. and the code itself can be changed to be compatible with any version of the cold staking. 
- Depend on monetary policy: the proposed formula is depending on the monetary policy, but if the new monetary policy is set now, even if it needs multiple hardforks the reward can be hardcoded following it. and this is in the case if you want a fully decentralized contract, in another case a simple function that can be called externally that set the block reward value following block.number will solve the issue.

@yograterol, can you please let us know about the new monetary policy.

Also another advantage that is not highlighted is the Cold Staking Treasury fund, none of the other proposals include how to distribute the fund in long term.

**Comment by yuriy77k** [comment]
I did not see any problem in that example. Alice get little more reward, because her part in first 40 blocks not 0.6666, but little bit more (she was only holder for 10 blocks). 
Bob get little bit more, because rewards pool was not 3200 CLO, but more on his part from first 10 blocks.

**Comment by yuriy77k** [comment]
I've added [protection against valuable manipulation of block.timestamp](https://github.com/yuriy77k/Cold-staking/blob/d8a8eddfa024770c63d37f344ee2dd775a59dbc2/ColdStaking.sol#L118-L125)

**Comment by RideSolo** [comment]
If you decide to go with @yuriy77k  's proposal, we should discuss the reward dependence to users transactions (withdraw, claim, stake).

The following statement is no true "Stakers can not directly affect each other with their deposits/withdrawals/claims. Reward calculation is more predictable" , check below for more information.

# Reward Computation dependent on Staker Actions (Transaction order) 

For better understading of the operation, The time of each block is set for 1 second. The following operation will be done by the stakers:

- Staker 1 stake  100 clo at block 0.
- Staker 2 stake  100 clo at block 3.
- Staker 3 stake 5000 clo at block 5.

the reward per block is set to be 100 clo, just for illustrative purpose.

The following table 1 shows the staking reward pool, the stakers weight and the total staking weight following the block number.

| Block             |  0    |   1   |   2   |   3   |   4   |   5   |   6  |   7  |
|-------------------|------:|------:|------:|------:|------:|------:|-----:|-----:|
| StakingRewardPool | 100   |  200  |  300  |  400  |  500  |  600  |  700 |  800 |
| staker 1 Weight   | 0     |  100  |  200  |  300  |  400  |  500  |  600 |  700 |
| staker 2 Weight   | 0     |    0  |    0  |    0  |  100  |  200  |  300 |  400 |
| staker 3 Weight   | 0     |    0  |    0  |    0  |    0  |    0  | 5000 | 10000|
| TotalStakingWeight| 0     |  100  |  200  |  300  |  500  |  700  | 5900 | 11100|
					Table 1

The following table 2 shows the stakers reward using data of table 1. 

| Block             |  0    |   1   |   2   |   3   |   4   |   5   |   6  |   7  | 
|-------------------|------:|------:|------:|------:|------:|------:|-----:|-----:|
| staker 1 reward   | 0     |  200  |  300  |  400  |  400  | ~428  |  ~71 |  ~50 | 
| staker 2 reward   | 0     |    0  |    0  |    0  |  100  | ~171  |  ~35 |  ~28 | 
| staker 3 reward   | 0     |    0  |    0  |    0  |    0  |    0  | ~593 | ~720 |
					Table 2
					
We can clearly see for example that the reward of staker 1 is dependent to others stakers interaction, and it is the same for every staker.

Another example is the following interactions that are done in order:
- Staker 3 stake 5000 clo.
- Staker 1 claim his reward after staker 3 stake. 
- Staker 3 withdraw his stake before the end of the month.
- Staker 2 claim after staker 3 withdrawal.
What will happen is staker 1 will get a smaller reward compared to staker 2, even if he staked before staker 2.
In production this use cases will happen all the time, since thousands of users will be claiming, withdrawing and staking.

**Comment by Dexaran** [comment]
@RideSolo 

>First of all, the proposal was to show an idea about how to handle the cold staking with a different angle so everyone can see that we have other possibility.

Completely agree that there is a wide range of possible implementations each with its own advantages and disadvantages. I think that the version that we are going to pick now may be replaced with a better one in future as well.

We are going to implement 5M32 monetary policy but it may change in future as well ([whitepaper](https://drive.google.com/file/d/18NsazbbcMwmas7TIZRbDslIfNCVeLgBY/view)).

>another advantage that is not highlighted is the Cold Staking Treasury fund, none of the other proposals include how to distribute the fund in long term.

I would say that Treasury Fund distribution is intended to be absolutely abstracted from Cold Staking mechanic. I know that this function exists in the Cold Stack contract, but I can not say that this is an advantage, and I can not say that it is necessary for us.

**Comment by Dexaran** [comment]
>The following statement is no true "Stakers can not directly affect each other with their deposits/withdrawals/claims. Reward calculation is more predictable"

Yes, the statement is not correct.
A staker can affect other stakers if he deposits a significant amount of funds in the contract before the others withdraw their rewards. Updated the comment.

>What will happen is staker 1 will get a smaller reward compared to staker 2, even if he staked before staker 2.

>In production this use cases will happen all the time, since thousands of users will be claiming, withdrawing and staking.

I assume that this is similar to the dispersion of mining rewards.

If a `miner_1` will start mining when the network hashrate is 10 Th/s but the network hashrate will grow to 20 Th/s then the reward of `miner_1` will be different from his initial expectations.

Hashrate is not constant because the miners come and go out. However, in large and stable networks (such are ETH or ETC), where there are a lot of miners, rewards are fairly predictable, because of a large number of participants (miners in this case) and the effect of unpredictable behavior of each participant is smoothed.

However, I admit that the parallel with the miners may not be accurate, since the stakers and miners are fundamentally different roles in the network.

**Comment by yuriy77k** [comment]
@RideSolo 

1. The reward is calculated and paid not for block, but for a month

2. This situation looks terrible if you look at it as presented (in blocks). But if interpellate it in real situation, we get:
30 days / 6 = 1 block is equivalent to 5 days
5000 CLO is about  5 times more then 900 CLO ( TotalStakingWeight). 
So if in Cold Stking doposits only 10% of total amount of CLO, then on 6 block (25 day) the  TotalStakingWeight = 10% *5 = 50%, than   amount equivalent 5000 must be more then 50% * 5 = 250% of total amount of CLO (which is impossible). 

3. In other side, if use RideSolo's variant, we get following:  

*Staker 1 stake 100 clo at block 1.*
*Staker 2 stake 5000 clo at block 4.*
| Block             |  0    |   1   |   2   |   3   |   4   |   5   |   6  |
|-------------------|------:|------:|------:|------:|------:|------:|-----:|
| StakingRewardPool | 100   |  100  |  100  |  100  |  100  |  100  |  100 |
| staker 1 Weight   | 0     |  100  |  100  |  100  |  100  |  100  |  100 |
| staker 2 Weight   | 0     |    0  |    0  |    0  |  5000 |  5000 | 5000 |
| staker 1 reward   | 0     |  100  |  200  |  300  | ~302  | ~304  | ~306 |
| staker 2 reward   | 0     |    0  |    0  |    0  |  ~98  | ~196  | ~294 |
					Table 1

So Staker 1 receive more reward with 100 CLO deposit, as Staker 2 with 5000 CLO deposit, but staking period of Staker 1 just in two times more then Staker 2 period. And what to do with unused StakingRewardPool for block 0?


4. Another situation which is possible with RideSolo's variant.

| Block             |  0    |   1   |   2   |   3   |   4   |   5   |   6  |
|-------------------|------:|------:|------:|------:|------:|------:|-----:|
| StakingRewardPool | 100   |  100  |  100  |  100  |  100  |  100  |  100 |
| staker 1 Weight   | 100   |  100  |  100  |  100  |  100  |  100  |  100 |
| staker 2 Weight   | 0     |  100  |  100  |  100  |  100  |  100  |  100 |
|     ....          | ...   |  ...  |  ...  |  ...  |  ...  |  ...  |  ... |
| staker 100 Weight | 0     |  100  |  100  |  100  |  100  |  100  |  100 |
| staker 1 reward   | 100   |  101  |  102  |  103  |  104  |  105  |  106 |
| staker 2 reward   | 0     |    1  |    2  |    3  |    4  |    5  |    6 |
|     ....          | ...   |  ...  |  ...  |  ...  |  ...  |  ...  |  ... |
| staker 100 reward | 0     |    1  |    2  |    3  |    4  |    5  |    6 |
					Table 2

So Staker 1 get much more rewards then others, who make deposit just on next block.

**Comment by RideSolo** [comment]
> A staker can affect other stakers if he deposits a significant amount of funds in the contract before the others withdraw their rewards. Updated the comment.
supposing a large number of starkers and not only one big staker, is similar since at the end they will have an accumulated weight, which will be a more common issue.

> I assume that this is similar to the dispersion of mining rewards.
> If a `miner_1` will start mining when the network hashrate is 10 Th/s but the network hashrate will grow to 20 Th/s then the reward of `miner_1` will be different from his initial expectations.
> Hashrate is not constant because the miners come and go out. However, in large and stable networks (such are ETH or ETC), where there are a lot of miners, rewards are fairly predictable, because of a large number of participants (miners in this case) and the effect of unpredictable behavior of each participant is smoothed.

It is not really the same case since miners previous rewards at block i will not be affected by the future hash rate of the network. in the opposite the future TotalStakingWeight will affect the previous reward at any given block or monthly based reward, since users do no stake all at the same time.

**Comment by RideSolo** [comment]
>     1. The reward is calculated and paid not for block, but for a month

Yes I understand this, but the month is not fixed for everyone, stakers can start staking when they want and the StakingRewardPool distribution doesn't take into account the cycles since everyone cycle is different. 

>     2. This situation looks terrible if you look at it as presented (in blocks). But if interpellate it in real situation, 

My point was to show that the reward is dependent on users actions. and as commented earlier you can think of that stakers as thousands of small stakers who comes in the last week of 1 staker 1 monthly round_interval.

**Comment by RideSolo** [comment]
@Dexaran @yuriy77k, correct If I'm wrong but the only way to give a fair reward and independent from any action to everyone is to fix a monthly cycle and with a fix reward pool for every month. for example all the starkers will be able to claim after every 30 days since cold staking start.

**Comment by RideSolo** [comment]
@Dexaran @yuriy77k, I didn't notice before but, my formula and yuriy's formula are exactly the same at the only difference that I use a round_interval of 1 block only.

**Comment by RideSolo** [comment]
@yuriy77k 

> My examples also show that **final** staker reward dependent on users actions even with RideSolo's formula. The difference is only in calculation methods, but staker would not estimate his final reward in any case.
> Therefore, I proposed a comparison with the deviants on shares. The company can work well during the year, and in the last month it incurs large losses, and shareholders will receive less revenue than they expected. And vice versa.
> Therefore, companies do not publish monthly or daily income per share, but do it once a year, according to the annual result.

The company give share holders their rewards in a fixed period from day x of the year to day y of the year, Yuriy's proposal set an x(i) and y(i) different for each staker i, the reward dependence comes from the non-fixed round_invetrval start of each user, as pointed out before the only way to avoid the reward dependence is to use the smalest unit of time available on the blockchain (round_interval = 1 block) so the stakers will be free to start staking at any given moment.

**Comment by yuriy77k** [comment]
> the reward dependence comes from the non-fixed round_invetrval start of each user, as pointed out before the only way to avoid the reward dependence is to use the smalest unit of time available on the blockchain (round_interval = 1 block) so the stakers will be free to start staking at any given moment.

round_invetrval  - a) time interval to froze Stakers deposits; b) time interval to calculate Stakers rewards.
In case of "a)" we can not make it = 1 block.


This is similar to the situation in real life: 
Alice goes to work for 5 days, but actually works only 1 hour a day. Bob went to work for 1 day, but worked 8 hours. How do I pay my salary?

a) by the days (Alice get more)
b) by the hour (Bob get more)

Both options are permissible, and both options have their advantages and disadvantages.

I think that it is not possible to make an ideal solution, without disadvantages, so we need to take one as a basis and work with it.

**Comment by yuriy77k** [comment]
> I think that the version that we are going to pick now may be replaced with a better one in future as well.

Replace Cold Staking contract is not trivial task. Will need to keep all customers address, amount of deposits and time of deposits... and others values...

**Comment by RideSolo** [comment]
> What do you think about calculation staker reward at fixed day / time? For example each 15 day of month. In this case every staker will receive rewards commensurate with his StakerWeight.

I do think that it will be the most fair solution, but some issues can be raised:

- Monthly network congestion, since all stakers will be claiming their rewards at the same period.
- And of course, what will be the repercusions on the coin price if a certain % of this claims will be sold.

> round_invetrval - a) time interval to froze Stakers deposits; b) time interval to calculate Stakers rewards.
> In case of "a)" we can not make it = 1 block.

Yes, I agree with you. how ever what I proposed is a round_interval=1 block to distribute the block reward (20% of the mining block reward) and a waiting time equal to 1 month to freeze the reward, where the reward is the total reward of every block.

> I think that it is not possible to make an ideal solution, without disadvantages, so we need to take one as a basis and work with it.

I agree with you, I will update my proposal and the code in few hours, to remove to disadvantages pointed out by Dexaran so we can decide.

**Comment by RideSolo** [comment]
> we can fix RewardPool on selected day, and staker can claim his reward on any day after (before next month Payouts day) of the first 

By not fixing a selected day, the stakers claim will be more distributed. this is part dexaran staking protocol.

> > And of course, what will be the repercusions on the coin price if a certain % of this claims will be sold.
> 
> This is true for any schema of payouts.

It is not true, since if the day is not fixed the claims will be more distributed. if there is any sell volume coming from the claim rewards the volume will be more distributed.

**Comment by yuriy77k** [comment]
> if the day is not fixed the claims will be more distributed. if there is any sell volume coming from the claim rewards the volume will be more distributed.
>> an x(i) and y(i) different for each staker i, the reward dependence comes from the non-fixed round_invetrval start of each user

so not fixed date - bad, fixed day - bad... :-)

**Comment by RideSolo** [comment]
> > if the day is not fixed the claims will be more distributed. if there is any sell volume coming from the claim rewards the volume will be more distributed.
> > > an x(i) and y(i) different for each staker i, the reward dependence comes from the non-fixed round_invetrval start of each user
> 
> so not fixed date - bad, fixed day - bad... :-)

My proposal solve this issue, the date is not fixed and in the same time the reward in independent.
since as explained 'I proposed is a round_interval=1 block to distribute the block reward (20% of the mining block reward) and a waiting time equal to 1 month to freeze the reward, where the reward is the total reward of every block.'

**Comment by yuriy77k** [comment]
> My proposal solve this issue, the date is not fixed and in the same time the reward in independent.
since as explained 'I proposed is a round_interval=1 block to distribute the block reward (20% of the mining block reward) and a waiting time equal to 1 month to freeze the reward, where the reward is the total reward of every block.'

1. How you get block reward?

2. If staker claim his rewards after 40 days, he get rewards amount for 30 or 40 days?

**Comment by RideSolo** [comment]
>     1. How you get block reward?

I have updated the proposal, the contract is completely independent from the monetary policy adopted or any future or changes through a planned or not planned hard fork. as I said i will share it soon.

>     2. If staker claim his rewards after 40 days, he get rewards amount for 30 or 40 days?

if a staker stake at time X and claim after 40 days. he will get a reward of 40 days. and his lastClaim time will be set to X+40days, which will be a good enhancement that will push the claims to be distributed through the time.

**Comment by RideSolo** [comment]
@Dexaran @yuriy77k , Please note that I have updated the proposal (below is the links to the code and the proposal)
- Proposal: https://github.com/RideSolo/Cold-staking/blob/master/proposal.pdf
- Code: https://github.com/RideSolo/Cold-staking/blob/master/ColdStaking.sol

Advantages: 
- Better reward calculation.
- Independent reward from users actions or transactions.
- Stakers can stake extra amount even after that they started staking (can be discussed), this feature will stop them from creating new address to stake more. Every time when user stake more the reward won't be given to him until 30 days from his last stake day which will make them wait more to be able to withdraw stake, if he doesn't wait and withdraw the reward will be redistributed.

Disadvantages: 
The disadvantages highlighted by Dexaran have been removed 
- Complexity: as shown [here](https://github.com/EthereumCommonwealth/Roadmap/issues/51#issuecomment-425654172) Yuriy's proposal formula and mine are almost the same.
- Dependence to monetary policy: Independent.
- Compatiblity with The First Stake: Compatible.

**Comment by Dexaran** [comment]
@RideSolo 
>Please note that I have updated the proposal

Reviewing...


>I do think that it will be the most fair solution, but some issues can be raised:

>Monthly network congestion, since all stakers will be claiming their rewards at the same period.
>And of course, what will be the repercusions on the coin price if a certain % of this claims will be sold.

Honestly, I don't think that this is a problem. A good opportunity [buy the dip] (https://www.investopedia.com/terms/b/buy-the-dips.asp) attracts the attention of traders (thus increasing the adoption, network effect, liquidity and potential community of bagholders), that is rather good than bad, in long-term. It is necessary that the time intervals were large enough (for example a month) that the price could recover.

**Comment by RideSolo** [comment]
> Honestly, I don't think that this is a problem.
> A good opportunity to [buy the dip](https://www.investopedia.com/terms/b/buy-the-dips.asp) attracts the attention of traders (thus increasing the adoption, network effect, liquidity and potential community of bagholders), that is rather good than bad, in long-term. It is necessary that the time intervals were large enough (for example a month) that the price could recover.
 
I got your point and it is true, but how about the network congestion we don't know the future usage of the cold staking contract or the adoption of Callisto cryptocurrency buy the crypto community.

**Comment by RideSolo** [comment]
@Dexaran  "large quantities of claim-transactions", since after each 15th of every month all the users will withdraw their reward.

**Comment by Dexaran** [comment]
Yes, this issue still stands and it is the disadvantage of having a fixed pay day.

Currently, the supply of CLO is 668M. Assuming that 70% of total CLOs are engaged in cold staking, Callisto bandwidth is 15 tx/s and an approximate amount of each stakers deposit is 1000 CLO, it will require **9 hours** to process all claim transactions.

In reality it will cause periodic increases in gas fees.

**Comment by yuriy77k** [comment]
> 3. At HardFork2 we change the address, which receives 20% of the block reward, from 0xaaaaaaaaaaaa to 0xbbbbbbbbbb. Thus, the new contract starts to receive income. Then users withdraw the remaining balance of the old contract until it will be equal to 0. New stakers continue staking with a new contract at address 0xbbbbbbbbbb.

But users need to wait for complete 30 days after deposit time. So money will be frozen, but without new rewards.

**Comment by RideSolo** [comment]
@Dexaran the proposed cold staking contract was updated, @yuriy77k  highlighted an error in the intervalReward computation.

**Comment by yuriy77k** [comment]
@Dexaran added functions `freez` and `withdraw_rewards`.
To complete CS contract, please, clarify following:

1. Do we allow staker to withdraw deposit before 30 days? (propose: No)

2. When staker claim reward after 40 days, he got reward for 30 (10 days left on his account and he can claim next time at least after 20 days) or 40 days? (purpose: 30 days. It will stimulate staker to hold longer.)

3. Set `max_delay` to 1 year or more? (propose: 2 years. The 1 year is very common interval.)

**Comment by Dexaran** [comment]
@yuriy77k 

Applied some changes https://github.com/yuriy77k/Cold-staking/pull/1 .

>Do we allow staker to withdraw deposit before 30 days? (propose: No)

No. We shouldn't allow stakers to claim before 30 days.

>When staker claim reward after 40 days, he got reward for 30 or 40 days? (propose: 30 days. It will stimulate staker to hold longer.)

30 days.

>Set max_delay to 1 year or more? (propose: 2 years. The 1 year is very common interval.)

Let's set it to 2 years then.

>Do we need the function claim_and_withdraw ?

This function was necessary in my version of the contract. Now it is not necessary because `withdraw_stake` claims reward automatically in this version. We can get rid of `claim_and_withdraw`.

>Do we need function reinvest which add all staker's reward to his deposit?

It is nice to have, however we can upgrade the contract in the future. Let's make the first version as simple as possible.

**Comment by yuriy77k** [comment]
>> Do we need function reinvest which add all staker's reward to his deposit?

> It is nice to have, however we can upgrade the contract in the future. Let's make the first version as simple as possible.

I already implement [reinvest](https://github.com/yuriy77k/Cold-staking/blob/5c4759b169f078719fe9c794ede1d59b50d8f383/ColdStaking.sol#L232-L258) function and it will not rise contract complexity too much.  However upgrade the contract in the future may cause more [inconvenience for stakers](https://github.com/EthereumCommonwealth/Roadmap/issues/51#issuecomment-426016471).

**Comment by Dexaran** [comment]
>I already implement reinvest function and it will not rise contract complexity too much.

Yes, I saw the code.
Each line introduces a new place for a bug to exist, even when each individual line is incredibly simple. Even when I fully understand the purpose of the new code, I know lurking in each line could be a subtle but important bug.

As we know, [writing bug-free code is almost impossible](https://en.wikibooks.org/wiki/Bug_Free_Programming). Let's recall the [80:20](https://dzone.com/articles/applying-8020-rule-software) rule as well. I'd like to cite this: `In summary, there is no doubt that focusing on the 20% of the features that give you 80% of the value will maximize the investment in software development and improve overall user satisfaction ... Therefore, reducing scope and not doing 100% of the features and functions is not only a valid strategy, but a prudent one.`

The main idea is that **we don't need 100% ideal smart-contract code right now, we need a perfectly working one**. It may only contain 70% - 80% of the "nice to have" features but it **must** be error-free. The cost of a single bug is too high.

It is better to have a working contract with some inconveniences and limited functionality than have a complicated and convenient but hacked contract.

Again, we are not developing an ideal implementation of the Cold Staking contract. We are developing the very first and experimental implementation which will be updated in the future.

I think that it must be as simple as possible.

>However upgrade the contract in the future may cause more inconvenience for stakers.

We will upgrade contract in the future. I think that everything will be fair if we announce the upgrading of the contract and let stakers know about it. They all will be in absolutely the same situation.

Anyways, [the inconvenience](https://github.com/EthereumCommonwealth/Roadmap/issues/51#issuecomment-426016471) that you pointed out make no sense compared to the disastrous consequences of a single bug in the contract.

Let me remind about [TheDAO hack](https://en.wikipedia.org/wiki/The_DAO_(organization)). [Parity Multisig hack 1](https://blog.zeppelin.solutions/on-the-parity-wallet-multisig-hack-405a8c12e8f7) and [Parity Multisig hack 2](https://medium.com/chain-cloud-company-blog/parity-multisig-hack-again-b46771eaa838) that I witnessed.

**Comment by yuriy77k** [comment]
> Anyways, the inconvenience that you pointed out make no sense compared to the disastrous consequences of a single bug in the contract.

I would agree with you if it be completely different function code, but code of `reinvest` is almost copy-paste from `claim` function + `start_staking` function. So if `reinvest` contain bug the above functions will also contain bug. And no risk to get new bug.

**Comment by Dexaran** [comment]
I tend to agree that my approach may cause a bit of inconvenience for the users, but it is normal for the very first release of the protocol.
It may also increase the number of transactions on the network, but I doubt that CLO will suffer serious bandwidth issues during the next half of the year.

Again, it is better to have a very primitive but working contract, than start with advanced hacked version.

**Comment by Dexaran** [comment]
I would suggest moving forward and make a decision on final values for constant variables (`staking_threshold`, `round_interval`, `max_delay` and so on).

After that, we should update the contract so that it will be ready for use without any changes. We will perform a security audit and a bug bounty on the version of the contract that will not undergo any changes before deployment.

**Comment by Dexaran** [comment]
### 1. staking_threshold

We have the following statement at the [Whitepaper](https://drive.google.com/file/d/18NsazbbcMwmas7TIZRbDslIfNCVeLgBY/view): `Minimal amount: No minimal threshold.`

I don't see any serious reason to change this. I think that it should be 0.
However, we need to keep the `staking_threshold` implementation at the contract because it is intended to be reusable. I'd like to propose Cold Staking protocol for ETC as well. If it will be decided to apply any minimum threshold for some reason, then it will be easier to do this with our current implementation.

### 2. round_interval

I'm proposing **27 days**.

Our initial implementation assumes ~1 month of locking time. Traders often make their decisions based on charts and candle closures. Therefore, it is better for them to have their coins available when the time to make a decision will come. That's why I'm proposing 27 days interval. 

If you deposit your funds at the beginning of a month then you will have the opportunity to claim it back right before the closure of the (1) week candle and (2) month candle at the same time.

![candle_close](https://user-images.githubusercontent.com/26142412/46502950-7ae0a080-c83a-11e8-8d21-60f9e9a95391.png)

### 3. max_delay

If you think that 2 years is a reasonable length then I have no objections. Let's use 2 years.

### 4. DateStartStaking

I'm proposing `1541894400`.

11/11/2018 @ 12:00am (UTC)

**Comment by yuriy77k** [comment]
> The result of the execution of reinvest() function is exactly the same as if a user withdrawn his staking reward and then deposited the withdrawal amount back with the second transaction.

Not exactly.  If staker hold 40 days and withdraw his deposit - he received reward only for 30 days (round_interval) and lose reward for 10 days. But reinvest() add full reward for 40 days to staker deposit (and begin new round), so it will be more profitable for users and stimulate it keep staking longer time.

**Comment by RideSolo** [comment]
@yuriy77k @Dexaran, I just want to highlight something with the current proposal.
The stakers should be pushed to claim their reward otherwise if we suppose a positive growth of the total staked amount, if a user wait two months for example to claim, his two months reward will be less that if he claims twice (the first and second month) basically the formula is not linear.

**Comment by Dexaran** [comment]
@yuriy77k 

>reinvest() add full reward for 40 days to staker deposit (and begin new round)

Yes, it is a bit more convenient for stakers.
If you re-stake each month or use the second account then it will still work properly.

I fully agree that the `reinvest()` function is a useful feature, but it is not of critical importance for us right now. That is why it is better to implement it after we make sure that the very basic version works properly.

>it will be more profitable for users and stimulate it keep staking longer time.

Honestly, I don't see how it incentivizes users to stake longer. Staking longer is not more profitable compared to the "withdraw & use another account" method.
I agree that it is a bit inconvenient but not critical. As the side effect, it can increase the number of transactions and active addresses of the Callisto network, but I don't think that it is a big problem.

@RideSolo 
>if a user wait two months for example to claim, his two months reward will be less that if he claims twice

I would say yes. We need to provide a description and highlight how the staking works. The description should be pinned somewhere so that everyone can find it easily.

**Comment by Dexaran** [comment]
# Final implementation pattern.

1. Security audit of the Cold Staking contract.
2 Bug bounty.
3. Contract will be compiled and deployed at Callisto Mainnet before the hardfork date.
4. At 11th November, the hardfork №1 changes will be enabled. Contract will start to receive 20% of block rewards. Staking will not be enabled instantly.
5. At 12th November 0:0 UTC the staking will become available.

Contract will be deployed at Callisto Testnet soon.

**Comment by Dexaran** [comment]
# Conclusion.

I've reviewed both Yuriy's and RideSolo's versions of the Cold Staking. 
RideSolo's version seems to be of arguable higher complexity. I think that for the first launch, the main focus is to show that [everything works as intended](https://github.com/EthereumCommonwealth/Roadmap/issues/51#issuecomment-427124840). Also, Callisto is an experimental project and both versions of the Staking Protocol could be a part of this experiment.

As the result, I've decided that the best option is to implement Yuriy's version at HF1 and test the RideSolo's version at this time. After that, the RideSolo's version of ColdStaking will be implemented at HF2.

In this way we will have the opportunity to test and compare both versions properly. However, I'd like to highlight that the most important aspect for the first implementation (HF1) is to ensure that everything works properly without a single bug.We must avoid such accidents like TheDAO hack by any costs.

**Comment by RideSolo** [comment]
> I've reviewed both Yuriy's and RideSolo's versions of the Cold Staking.
> RideSolo's version seems to be of arguable higher complexity. I think that for the first launch, the main focus is to show that [everything works as intended](https://github.com/EthereumCommonwealth/Roadmap/issues/51#issuecomment-427124840). Also, Callisto is an experimental project and both versions of the Staking Protocol could be a part of this experiment.

The most important for me was to show the inconvenients of the yuriy's proposal, so everyone will know what to expect, and we all know that any possible idea has inconvenient and advantages.

> As the result, I've decided that the best option is to implement Yuriy's version at HF1 and test the RideSolo's version at this time. After that, the RideSolo's version of ColdStaking will be implemented at HF2.

Personally I will keep the repository up to date until then.

> In this way we will have the opportunity to test and compare both versions properly. However, I'd like to highlight that the most important aspect for the first implementation (HF1) is to ensure that everything works properly without a single bug.We must avoid such accidents like TheDAO hack by any costs.

I agree with your vision.

**Comment by Dexaran** [comment]
> I do think that it is better to set the withdrawal address to the actual Callisto Staking Reserve 

Sounds reasonable.

**Comment by Geo25rey** [comment]
@Dexaran Sorry, I'm late to the party, but I'm curious to why a staking contract can't have funds added to it before the end of 27 days. Surely, there can be a way to allow a user to continuously add to their stake. Maybe allow each wallet to have multiple staking contracts that can start at different times?

Another solution would be allowing certain modifications (e.g. adding funds) of a staking contract and whenever there is such a modification, the round will be reset while adding on the portion of the staking contract that has already been established to the replacement staking contract. This assumes that the rounds are local to each staking contract.
